packages/pwa-kit-runtime/src/ssr/server/build-remote-server.js

```diff
@@ -935,6 +938,97 @@ export const RemoteServerFactory = {
+        'connect-src': ["'self'", 'api.cquotient.com', runtimeAdmin],
+        'frame-ancestors': [runtimeAdmin],
+        'img-src': ["'self'", '*.commercecloud.salesforce.com', 'data:'],
+        'script-src': ["'self'", "'unsafe-eval'", runtimeAdmin]
+    }
+
+    const setHeader = res.setHeader
+    res.setHeader = (name, value) => {
+        let modifiedValue = value
+        switch (name?.toLowerCase()) {
+            case CONTENT_SECURITY_POLICY: {
+                // If multiple Content-Security-Policy headers are provided, then the most restrictive
+                // option is chosen for each directive. Therefore, we must modify *all* directives to
+                // ensure that our required directives will work as expected.
+                // Ref: https://w3c.github.io/webappsec-csp/#multiple-policies
+                modifiedValue = Array.isArray(value)
+                    ? value.map((item) => addRequiredDirectives(item, directives))
+                    : addRequiredDirectives(value, directives)
+                break
+            }
+            case STRICT_TRANSPORT_SECURITY: {
+                // Block setting this header on local development server - it will break things!
+                if (!isRemote()) return
+                break
+            }
+            default: {
+                break
+            }
+        }
+        return setHeader.call(res, name, modifiedValue)
+    }
+    // Provide an initial CSP (or patch the existing header)
+    res.setHeader(CONTENT_SECURITY_POLICY, res.getHeader(CONTENT_SECURITY_POLICY) ?? '')
+    // Provide an initial value for HSTS, if not already set - use default from `helmet`
+    if (!res.hasHeader(STRICT_TRANSPORT_SECURITY)) {
+        res.setHeader(STRICT_TRANSPORT_SECURITY, 'max-age=15552000; includeSubDomains')
+    }
+    next()
+}
+
+/**
+ * Updates the given Content-Security-Policy header to include all directives required by PWA Kit.
+ * @param {string} original Original Content-Security-Policy header
+ * @returns {string} Modified Content-Security-Policy header
+ * @private
+ */
+const addRequiredDirectives = (original, required) => {
+    const directives = original
+        .trim()
+        .split(';')
+        .reduce((acc, directive) => {
+            const text = directive.trim()
+            if (text) {
+                const [name, ...values] = text.split(/ +/)
+                acc[name] = values
+            }
+            return acc
+        }, {})
+    // Add missing required CSP directives
+    for (const [name, values] of Object.entries(required)) {
+        // Wrapping with `[...new Set(array)]` removes duplicate entries
+        directives[name] = [...new Set([...(directives[name] ?? []), ...values])]
+    }
+    // Always upgrade insecure requests when deployed, never upgrade on local dev server
+    if (isRemote()) {
+        directives['upgrade-insecure-requests'] = []
+    } else {
+        delete directives['upgrade-insecure-requests']
+    }
```

Comment:
Only comment I could think of is for this function (which is otherwise nice) is that this section is a little bit of the black sheep, if we renamed the function it would be `addRequiredDirectivesAndMabyeInsecureRequests`. Have you thought of including conditionally modifying the required directives before calling this function? You can always stripe out the "null" value directives if you are concerned with those, since I know we have/had them in our directive definitions. 

Thoughts?